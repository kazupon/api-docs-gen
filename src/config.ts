import type { Resolver, Processor } from './types'

/**
 * Configration
 */
export interface Config {
  /**
   * markdown link reference resolver
   */
  resolver?: Resolver
  /**
   * processor
   */
  processor: Processor
}
