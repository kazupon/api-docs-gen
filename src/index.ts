export { getSafePathFromDisplayName, escapeText } from './utils'
export * from './builder'
export * from './config'
export {
  multi as multiProcessor,
  toc as tocProcessor,
  getCustomTags,
  getDocSectionContent
} from './processor'
export { multi as multiResolver, toc as tocResolver } from './resolver'
export { generate } from './generator'
