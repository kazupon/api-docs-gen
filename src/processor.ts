import {
  ApiItemKind,
  ApiModel,
  ApiPackage,
  ApiItem,
  ApiDeclaredItem,
  ApiParameterListMixin,
  IResolveDeclarationReferenceResult,
  ApiEnumMember
} from '@microsoft/api-extractor-model'
import {
  DocBlock,
  DocNodeKind,
  DocSection,
  DocPlainText,
  DocCodeSpan,
  IDocLinkTagParameters,
  DocDeclarationReference,
  DocFencedCode
} from '@microsoft/tsdoc'
import type { DeclarationReference } from '@microsoft/tsdoc/lib/beta/DeclarationReference'
import { debug as Debug } from 'debug'
import type {
  MarkdownContent,
  ReferenceResolver,
  GenerateStyle
} from './config'
import { ContentBuilder, createContentBuilder } from './builder'
import { escapeText } from './utils'

const debug = Debug('api-docs-gen:processor')

export function process(
  model: ApiModel,
  pkg: ApiPackage,
  style: GenerateStyle,
  resolver: ReferenceResolver
): string | MarkdownContent[] {
  // build
  function build(pkg: ApiPackage): Map<string, ContentBuilder> {
    const builders = new Map<string, ContentBuilder>()
    const items = pkg.members[0] ? pkg.members[0].members : []

    buildMarkdown(builders, items)
    return builders
  }

  // build markdown
  function buildMarkdown(
    builders: Map<string, ContentBuilder>,
    items: readonly ApiItem[]
  ): void {
    for (const item of items) {
      // console.log('buildMarkdown item', item)
      const { kind } = item
      let builder = builders.get(kind)
      if (!builder) {
        builder = createContentBuilder()
        builder.pushline(`# ${kind}`)
        builder.newline()
        builders.set(kind, builder)
      }

      switch (kind) {
        case ApiItemKind.Function:
          buildFunctionMarkdown(builder, item)
          break
        case ApiItemKind.Enum:
          buildEnumMarkdown(builder, item)
          break
        case ApiItemKind.Interface:
          buildInterfaceMarkdown(builder, item)
          break
        case ApiItemKind.Class:
          buildClassMarkdown(builder, item)
          break
        case ApiItemKind.TypeAlias:
          buildTypeAliasMarkdown(builder, item)
          break
        case ApiItemKind.Variable:
          buildVariableMarkdown(builder, item)
          break
        default:
          break
      }
    }
  }

  // build function markdown
  function buildFunctionMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // parameters
    const itemParam = item as ApiParameterListMixin
    if (itemParam.parameters) {
      builder.pushline(`### Parameters`)
      builder.newline()
      builder.pushline(`| Parameter | Type | Description |`)
      builder.pushline(`| --- | --- | --- |`)
      for (const p of itemParam.parameters) {
        builder.pushline(
          `| ${p.name} | ${p.parameterTypeExcerpt.text.trim()} | ${
            p.tsdocParamBlock && p.tsdocParamBlock.content
              ? getDocSectionContent(
                  model,
                  pkg,
                  // @ts-ignore TODO:
                  p.tsdocParamBlock.content,
                  item,
                  style,
                  resolver
                )
              : ''
          } |`
        )
      }
      builder.newline()
    }

    // returns
    if (docs.returnsBlock) {
      builder.pushline(`### Returns`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.returnsBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // throws
    // @ts-ignore TODO:
    const throws = getCustomTags(docs.customBlocks, '@throws')
    if (throws.length > 0) {
      builder.pushline(`### Throws`)
      builder.newline()
      for (const t of throws) {
        let text = `${getDocSectionContent(
          model,
          pkg,
          t.content,
          item,
          style,
          resolver
        )}`
        if (throws.length > 1) {
          text = `- ` + text
        }
        builder.pushline(text)
      }
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  // build enum markdown
  function buildEnumMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // members
    if (item.members) {
      builder.pushline(`### Members`)
      builder.newline()
      builder.pushline(`| Member | Value| Description |`)
      builder.pushline(`| --- | --- | --- |`)
      for (const m of item.members) {
        const memberEnum = m as ApiEnumMember
        const memberDeclared = m as ApiDeclaredItem
        builder.pushline(
          `| ${memberEnum.displayName} | ${
            memberEnum.initializerExcerpt.text
          } | ${
            memberDeclared.tsdocComment &&
            memberDeclared.tsdocComment.summarySection
              ? getDocSectionContent(
                  model,
                  pkg,
                  // @ts-ignore TODO:
                  memberDeclared.tsdocComment.summarySection,
                  item,
                  style,
                  resolver
                )
              : ''
          } |`
        )
      }
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  function buildMarkdownForClassinizable(
    builder: ContentBuilder,
    item: ApiItem,
    type: string
  ) {
    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // parameters
    const itemParam = item as ApiParameterListMixin
    if ((type === 'constrcutor' || type === 'method') && itemParam.parameters) {
      builder.pushline(`*Parameters*`)
      builder.newline()
      builder.pushline(`| Parameter | Type | Description |`)
      builder.pushline(`| --- | --- | --- |`)
      for (const p of itemParam.parameters) {
        builder.pushline(
          `| ${p.name} | ${p.parameterTypeExcerpt.text.trim()} | ${
            p.tsdocParamBlock && p.tsdocParamBlock.content
              ? getDocSectionContent(
                  model,
                  pkg,
                  // @ts-ignore TODO:
                  p.tsdocParamBlock.content,
                  item,
                  style,
                  resolver
                )
              : ''
          } |`
        )
      }
      builder.newline()
    }

    // returns
    if (type === 'method' && docs.returnsBlock) {
      builder.pushline(`### Returns`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.returnsBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // throws
    // @ts-ignore TODO:
    const throws = getCustomTags(docs.customBlocks, '@throws')
    if ((type === 'constructor' || type === 'method') && throws.length > 0) {
      builder.pushline(`### Throws`)
      builder.newline()
      for (const t of throws) {
        let text = `${getDocSectionContent(
          model,
          pkg,
          t.content,
          item,
          style,
          resolver
        )}`
        if (throws.length > 1) {
          text = `- ` + text
        }
        builder.pushline(text)
      }
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  // build interface markdown
  function buildInterfaceMarkdown(
    builder: ContentBuilder,
    item: ApiItem
  ): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()
    buildMarkdownForClassinizable(builder, item, 'interface')
    builder.newline()

    const methods = item.members.filter(m => m.kind === 'MethodSignature')
    builder.pushline(`### Methods`)
    builder.newline()
    for (const method of methods) {
      builder.pushline(`#### ${method.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, method, 'method')
    }
    builder.newline()

    const properties = item.members.filter(m => m.kind === 'PropertySignature')
    builder.pushline(`### Properties`)
    builder.newline()
    for (const property of properties) {
      builder.pushline(`#### ${property.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, property, 'property')
    }
    builder.newline()
  }

  // build class markdown
  function buildClassMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()
    buildMarkdownForClassinizable(builder, item, 'class')
    builder.newline()

    const [ctor] = item.members.filter(m => m.kind === 'Constructor')
    builder.pushline(`### Constructor`)
    builder.newline()
    buildMarkdownForClassinizable(builder, ctor, 'constructor')
    builder.newline()

    const methods = item.members.filter(m => m.kind === 'Method')
    builder.pushline(`### Methods`)
    builder.newline()
    for (const method of methods) {
      builder.pushline(`#### ${method.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, method, 'method')
    }
    builder.newline()

    const properties = item.members.filter(m => m.kind === 'Property')
    builder.pushline(`### Properties`)
    builder.newline()
    for (const property of properties) {
      builder.pushline(`#### ${property.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, property, 'property')
    }
    builder.newline()
  }

  // build type alias markdown
  function buildTypeAliasMarkdown(
    builder: ContentBuilder,
    item: ApiItem
  ): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  // build variable markdown
  function buildVariableMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  function generate(builders: Map<string, ContentBuilder>): MarkdownContent[] {
    const contents = []
    for (const [kind, builder] of builders) {
      const content = builder.content
      contents.push({
        filename: `${kind.toLowerCase()}.md`,
        body: content
      })
    }
    return contents
  }

  // build markdown content!
  const builders = build(pkg)

  // generate markdown content!
  return generate(builders)
}

export function getDocSectionContent(
  model: ApiModel,
  pkg: ApiPackage,
  content: DocSection,
  contextItem: ApiItem,
  style: GenerateStyle,
  resolver: ReferenceResolver
): string {
  let ret = ''

  for (const n of content.nodes) {
    if (n.kind === DocNodeKind.Paragraph) {
      for (const nn of n.getChildNodes()) {
        if (nn.kind === DocNodeKind.PlainText) {
          ret += ((nn as unknown) as DocPlainText).text
        } else if (nn.kind === DocNodeKind.CodeSpan) {
          ret += `\`${((nn as unknown) as DocCodeSpan).code}\``
        } else if (nn.kind === DocNodeKind.LinkTag) {
          const {
            linkText,
            urlDestination,
            codeDestination
          } = (nn as unknown) as IDocLinkTagParameters
          if (codeDestination) {
            const result = resolveDeclarationReference(
              model,
              codeDestination,
              contextItem
            )
            if (result.resolvedApiItem) {
              const filepath = resolver(
                style,
                result.resolvedApiItem,
                model,
                pkg
              )
              if (linkText) {
                const encodedLinkText = escapeText(
                  linkText.replace(/\s+/g, ' ')
                )
                ret += `[${encodedLinkText}](${filepath})`
              } else {
                ret += `[${result.resolvedApiItem.displayName}](${filepath})`
              }
            }
          } else {
            if (linkText) {
              if (urlDestination) {
                const encodedLinkText = escapeText(
                  linkText.replace(/\s+/g, ' ')
                )
                ret += `[${encodedLinkText}](${urlDestination})`
              } else {
                ret += escapeText(linkText.replace(/\s+/g, ' '))
              }
            } else {
              if (urlDestination) {
                ret += urlDestination
              }
            }
          }
        }
      }
    } else if (n.kind === DocNodeKind.FencedCode) {
      const fenced = (n as unknown) as DocFencedCode
      ret += `\n`
      ret += `\`\`\`${fenced.language}\n`
      ret += fenced.code
      ret += `\`\`\``
    } else {
      debug(`ignore build doc section: ${n.kind}`)
    }
  }

  return ret
}

export function getCustomTags(
  customBlocks: readonly DocBlock[],
  tag: string
): DocBlock[] {
  return customBlocks.filter(x => x.blockTag.tagName === tag)
}

function resolveDeclarationReference(
  model: ApiModel,
  reference: DocDeclarationReference | DeclarationReference,
  item: ApiItem
): IResolveDeclarationReferenceResult {
  // @ts-ignore TODO:
  return model.resolveDeclarationReference(reference, item)
}
