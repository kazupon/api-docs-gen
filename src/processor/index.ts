export { process as multi } from './multi'
export { process as toc } from './toc'
export * from './utils'
