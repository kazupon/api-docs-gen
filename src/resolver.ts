import type { ApiItem, ApiModel } from '@microsoft/api-extractor-model'
import { getSafePathFromDisplayName } from './utils'

export function resolve(item: ApiItem, model: ApiModel): string {
  if (item.kind === 'Model') {
    return './index'
  }

  let baseName = ''
  for (const hierarchyItem of item.getHierarchy()) {
    const qualifiedName = getSafePathFromDisplayName(hierarchyItem.displayName)
    switch (hierarchyItem.kind) {
      case 'Model':
      case 'EntryPoint':
      case 'Package':
        break
      case 'Enum':
      case 'Function':
      case 'Variable':
      case 'TypeAlias':
        baseName += `${hierarchyItem.kind.toLowerCase()}#${qualifiedName}`
        break
      default:
        baseName += '.' + qualifiedName
    }
  }
  return './' + baseName
}
