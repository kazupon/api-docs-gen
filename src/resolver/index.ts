export { resolve as multi } from './multi'
export { resolve as toc } from './toc'
