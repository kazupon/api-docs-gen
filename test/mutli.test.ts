import path from 'path'
import { ApiModel } from '@microsoft/api-extractor-model'
import { multi as multiResolver } from '../src/resolver'
import { multi as multiProcessor } from '../src/processor'
import { loadTSDocConfig, mergeTSDocTagDefinition } from '../src/tsdoc'
import { MarkdownContent, GenerateStyle } from '../src/config'

test('multi', () => {
  const apiModel = new ApiModel()
  const target = path.resolve(__dirname, './fixtures/library1.api.json')
  const apiPackage = apiModel.loadPackage(target)
  const contents = multiProcessor(
    apiModel,
    apiPackage,
    GenerateStyle.Prefix,
    multiResolver
  ) as MarkdownContent[]
  for (const content of contents) {
    expect(content.body).toMatchSnapshot(content.filename)
  }
})

test('multi with custom tags', () => {
  const tsdocConfigPath = path.resolve(
    __dirname,
    './fixtures/tsdoc.library1.json'
  )
  const { tagDefinitions } = loadTSDocConfig(tsdocConfigPath)
  const customTags = mergeTSDocTagDefinition(tagDefinitions)
  const apiModel = new ApiModel()
  const target = path.resolve(__dirname, './fixtures/library1.custom.api.json')
  const apiPackage = apiModel.loadPackage(target)
  const contents = multiProcessor(
    apiModel,
    apiPackage,
    GenerateStyle.Prefix,
    multiResolver,
    customTags
  ) as MarkdownContent[]
  for (const content of contents) {
    expect(content.body).toMatchSnapshot(content.filename)
  }
})
