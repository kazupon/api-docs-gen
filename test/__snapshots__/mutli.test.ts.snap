// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multi with custom tags: class.md 1`] = `
"# Class

## Calculator

Calculator class

**Signature:**
\`\`\`typescript
export declare class Calculator implements Calculatable 
\`\`\`

### Remarks

This is remarks of Calculator class

### Examples


\`\`\`javascript
const c = new Calculator()
const v1 = c.add(1, 1)
const v2 = c.sub(1, 1)
\`\`\`



### Constructor

Conssutructor of usage

**Signature:**
\`\`\`typescript
constructor(type: string);
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| type | string | calculator type |


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |

#### sub

sub method

**Signature:**
\`\`\`typescript
sub(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`

#### type

calcurator types

 'simple'

**Signature:**
\`\`\`typescript
type: string;
\`\`\`


"
`;

exports[`multi with custom tags: enum.md 1`] = `
"# Enum

## ErrorCodes

Error Code

**Signature:**
\`\`\`typescript
export declare enum ErrorCodes 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| InvalidFormat | 1 | Invalid format |
| Succcess | 0 | Success |

## TokenChars

Token Characters

**Signature:**
\`\`\`typescript
export declare enum TokenChars 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| Modulo | \\"%\\" | Modulo charactor |
| Plus | \\"+\\" | Plus charactor |

### Remarks

This is remarks of Token Chararaceters

"
`;

exports[`multi with custom tags: function.md 1`] = `
"# Function

## add

add function : \`x\`

**Signature:**
\`\`\`typescript
export declare function add(a: number, b: number): number;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target \`1\` |
| b | number | target 2 |

### Returns

 result as \`a\` + \`b\`

### Throws

- \`SyntaxError\` this is syntax error
- \`Error\` this is general error

### Remarks

This is add function remarks: [ErrorCodes](./library1-enum#errorcodes). See [add](https://foo.bar.com/api/add).

### Examples

example of \`add\` function:
\`\`\`javascript
console.log(add(1, 1))
\`\`\`


## undumbify

un dummy

**Signature:**
\`\`\`typescript
export declare function undumbify<T>(dummy: DumbType<T>): T;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| dummy | DumbType&lt;T&gt; | A dummy |

"
`;

exports[`multi with custom tags: interface.md 1`] = `
"# Interface

## Calculatable

Calculatable interface

**Signature:**
\`\`\`typescript
export interface Calculatable 
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`


"
`;

exports[`multi with custom tags: typealias.md 1`] = `
"# TypeAlias

## A

A

**Signature:**
\`\`\`typescript
export declare type A = string;
\`\`\`

## B

[A](./library1-typealias#a)

**Signature:**
\`\`\`typescript
export declare type B = number;
\`\`\`

## DumbType

A dumb type

**Signature:**
\`\`\`typescript
export declare type DumbType<T> = {
    foo: T;
};
\`\`\`

## FallbackLocale

Falback Locale

**Signature:**
\`\`\`typescript
export declare type FallbackLocale = Locale | Locale[] | {
    [locale in string]: Locale[];
} | false;
\`\`\`

### Remarks

This is remarks of \`Fallback Locale\`

## Locale

Locale

**Signature:**
\`\`\`typescript
export declare type Locale = string;
\`\`\`

## LocaleMessage

Locale Message resources

**Signature:**
\`\`\`typescript
export declare type LocaleMessage = string | LocaleMessage[];
\`\`\`

"
`;

exports[`multi with custom tags: variable.md 1`] = `
"# Variable

## Config

Configrations

 

**Signature:**
\`\`\`typescript
Config: {
    [name: string]: unknown;
}
\`\`\`

## VERSION

app version

**Signature:**
\`\`\`typescript
VERSION = \\"1.0.0\\"
\`\`\`

### Remarks

you can lookup application version that is semver format.

### Examples


\`\`\`javascript
console.log(VERSION)
\`\`\`


"
`;

exports[`multi: class.md 1`] = `
"# Class

## Calculator

Calculator class

**Signature:**
\`\`\`typescript
export declare class Calculator implements Calculatable 
\`\`\`

### Remarks

This is remarks of Calculator class

### Examples


\`\`\`javascript
const c = new Calculator()
const v1 = c.add(1, 1)
const v2 = c.sub(1, 1)
\`\`\`



### Constructor

Conssutructor of usage

**Signature:**
\`\`\`typescript
constructor(type: string);
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| type | string | calculator type |


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |

#### sub

sub method

**Signature:**
\`\`\`typescript
sub(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`

#### type

calcurator types

**Signature:**
\`\`\`typescript
type: string;
\`\`\`


"
`;

exports[`multi: enum.md 1`] = `
"# Enum

## ErrorCodes

Error Code

**Signature:**
\`\`\`typescript
export declare enum ErrorCodes 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| InvalidFormat | 1 | Invalid format |
| Succcess | 0 | Success |

## TokenChars

Token Characters

**Signature:**
\`\`\`typescript
export declare enum TokenChars 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| Modulo | \\"%\\" | Modulo charactor |
| Plus | \\"+\\" | Plus charactor |

### Remarks

This is remarks of Token Chararaceters

"
`;

exports[`multi: function.md 1`] = `
"# Function

## add

add function : \`x\`

**Signature:**
\`\`\`typescript
export declare function add(a: number, b: number): number;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target \`1\` |
| b | number | target 2 |

### Returns

 result as \`a\` + \`b\`

### Throws

- \`SyntaxError\` this is syntax error
- \`Error\` this is general error

### Remarks

This is add function remarks: [ErrorCodes](./library1-enum#errorcodes).

See [add](https://foo.bar.com/api/add).

### Examples

example of \`add\` function:
\`\`\`javascript
console.log(add(1, 1))
\`\`\`


## undumbify

un dummy

**Signature:**
\`\`\`typescript
export declare function undumbify<T>(dummy: DumbType<T>): T;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| dummy | DumbType&lt;T&gt; | A dummy |

"
`;

exports[`multi: interface.md 1`] = `
"# Interface

## AddCallable

Add Callable interface

**Signature:**
\`\`\`typescript
export interface AddCallable 
\`\`\`


### Functions

#### &lt;A extends number = number, B extends number = number&gt;(a: A, b: B): number;

add callable object

**Signature:**
\`\`\`typescript
<A extends number = number, B extends number = number>(a: A, b: B): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | A | target 1 |
| b | B | target 2 |

### Returns

 added value

#### &lt;A extends number = number, B extends number = number&gt;(a: A, b: B): string;

add callable object

**Signature:**
\`\`\`typescript
<A extends number = number, B extends number = number>(a: A, b: B): string;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | A | target 1 |
| b | B | target 2 |

### Returns

 added string value


## Calculatable

Calculatable interface

**Signature:**
\`\`\`typescript
export interface Calculatable 
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |

### Returns

 added value


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`


## MyOptions

My Options

**Signature:**
\`\`\`typescript
export interface MyOptions 
\`\`\`


### Properties

#### field

Field

**Signature:**
\`\`\`typescript
field?: string;
\`\`\`


"
`;

exports[`multi: typealias.md 1`] = `
"# TypeAlias

## A

A

**Signature:**
\`\`\`typescript
export declare type A = string;
\`\`\`

## B

[A](./library1-typealias#a)

**Signature:**
\`\`\`typescript
export declare type B = number;
\`\`\`

## DumbType

A dumb type

**Signature:**
\`\`\`typescript
export declare type DumbType<T> = {
    foo: T;
};
\`\`\`

## FallbackLocale

Falback Locale

**Signature:**
\`\`\`typescript
export declare type FallbackLocale = Locale | Locale[] | {
    [locale in string]: Locale[];
} | false;
\`\`\`

### Remarks

This is remarks of \`Fallback Locale\`

## Locale

Locale

**Signature:**
\`\`\`typescript
export declare type Locale = string;
\`\`\`

## LocaleMessage

Locale Message resources

**Signature:**
\`\`\`typescript
export declare type LocaleMessage = string | LocaleMessage[];
\`\`\`

"
`;

exports[`multi: variable.md 1`] = `
"# Variable

## Config

Configrations

**Signature:**
\`\`\`typescript
Config: {
    [name: string]: unknown;
}
\`\`\`

## VERSION

app version

**Signature:**
\`\`\`typescript
VERSION = \\"1.0.0\\"
\`\`\`

### Remarks

you can lookup application version that is semver format.

### Examples


\`\`\`javascript
console.log(VERSION)
\`\`\`


"
`;
